Shard Operators.cpp

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #62188
* #62187
* #62186
* #62278
* __->__ #62185
* #62184

This file can take 5 minutes on its own to compile, and is the single limiting
factor for compile time of `libtorch_cpu` on a 32-core threadripper. Instead,
sharding into 5 files that take around 1 minute each cuts a full minute off the
overall build time.

This also factors out the `.findSchemaOrThrow(...).typed` step so the code can
be shared between `call` and `redispatch`.

Differential Revision: [D29962049](https://our.internmc.facebook.com/intern/diff/D29962049)

peterbell10 committed:
Shard Operators.cpp

This file can take 5 minutes on its own to compile, and is the single limiting
factor for compile time of `libtorch_cpu` on a 32-core threadripper. Instead,
sharding into 5 files that take around 1 minute each cuts a full minute off the
overall build time.

This also factors out the `.findSchemaOrThrow(...).typed` step so the code can
be shared between `call` and `redispatch`.

[ghstack-poisoned]

peterbell10 committed:
Update on "Shard Operators.cpp"


This file can take 5 minutes on its own to compile, and is the single limiting
factor for compile time of `libtorch_cpu` on a 32-core threadripper. Instead,
sharding into 5 files that take around 1 minute each cuts a full minute off the
overall build time.

This also factors out the `.findSchemaOrThrow(...).typed` step so the code can
be shared between `call` and `redispatch`.

[ghstack-poisoned]

peterbell10 committed:
Update on "Shard Operators.cpp"


This file can take 5 minutes on its own to compile, and is the single limiting
factor for compile time of `libtorch_cpu` on a 32-core threadripper. Instead,
sharding into 5 files that take around 1 minute each cuts a full minute off the
overall build time.

This also factors out the `.findSchemaOrThrow(...).typed` step so the code can
be shared between `call` and `redispatch`.

[ghstack-poisoned]

peterbell10 committed:
Update on "Shard Operators.cpp"


This file can take 5 minutes on its own to compile, and is the single limiting
factor for compile time of `libtorch_cpu` on a 32-core threadripper. Instead,
sharding into 5 files that take around 1 minute each cuts a full minute off the
overall build time.

This also factors out the `.findSchemaOrThrow(...).typed` step so the code can
be shared between `call` and `redispatch`.

[ghstack-poisoned]

peterbell10 committed:
Update on "Shard Operators.cpp"


This file can take 5 minutes on its own to compile, and is the single limiting
factor for compile time of `libtorch_cpu` on a 32-core threadripper. Instead,
sharding into 5 files that take around 1 minute each cuts a full minute off the
overall build time.

This also factors out the `.findSchemaOrThrow(...).typed` step so the code can
be shared between `call` and `redispatch`.

[ghstack-poisoned]

ezyang commented on tools/codegen/gen.py:
```diff
@@ -276,9 +284,7 @@ def __call__(self, f: NativeFunction) -> Optional[str]:
-    static auto op = c10::Dispatcher::singleton()
-        .findSchemaOrThrow("aten::{f.func.name.name}", "{f.func.name.overload_name}")
-        .typed<{sig.type()}>();
+    static auto op = create_{name}_typed_handle();
```
Regarding the change here, would it be possible to make use of the ATen Operators API to do the handle creation here? If the typed handle function doesn't exist on the class, you would add it to the class (rather than having a new top level function which you have to namespace, etc.)

peterbell10 replied:
I kept it in the source file to avoid additional header includes in `Operators.h`. IMO this is fine because all you can do with a `TypedOperatorHandle` is `call` and `redispatch` so at that point you should just use the static functions, right?

bdhirsh commented on tools/codegen/gen.py:
```diff
@@ -1169,10 +1175,17 @@ def make_file_manager(install_dir: str) -> FileManager:
+    def key_func(fn: NativeFunction) -> str:
+        return fn.func.name.unambiguous_name()
+
+    cpu_fm.write_sharded(
```
It might be marginally useful to have an `OperatorsEverything.cpp` file that doesn't get built but doesn't force you to grep for the correct shard when you want to look at metadata about a particular operator. the sharding for autograd codegen does something similar with `VariableTypeEverything.cpp` here: https://github.com/pytorch/pytorch/pull/62063

peterbell10 replied:
Take a look at the first PR in this stack. The logic for `FooEverything.cpp` is all there in `write_sharded` :)

bdhirsh replied:
didn't notice that on the first pass, nice :)

peterbell10 commented:
I think this test failure is real:
```
Jul 26 21:56:31 /var/lib/jenkins/workspace/tools/code_analyzer/run_analyzer.sh: line 26: 13121 Segmentation fault      (core dumped)
```
I'm not 100% but I think the code analyser might expect `findSchemaOrThrow` to have literal arguments? If that sounds right I can move back to code-generated string literal for now.

ezyang commented:
> I'm not 100% but I think the code analyser might expect findSchemaOrThrow to have literal arguments? If that sounds right I can move back to code-generated string literal for now.

Yes, that sounds familiar. So you'll have to revert it alas.

peterbell10 committed:
Update on "Shard Operators.cpp"


This file can take 5 minutes on its own to compile, and is the single limiting
factor for compile time of `libtorch_cpu` on a 32-core threadripper. Instead,
sharding into 5 files that take around 1 minute each cuts a full minute off the
overall build time.

This also factors out the `.findSchemaOrThrow(...).typed` step so the code can
be shared between `call` and `redispatch`.

[ghstack-poisoned]

peterbell10 committed:
Update on "Shard Operators.cpp"


This file can take 5 minutes on its own to compile, and is the single limiting
factor for compile time of `libtorch_cpu` on a 32-core threadripper. Instead,
sharding into 5 files that take around 1 minute each cuts a full minute off the
overall build time.

This also factors out the `.findSchemaOrThrow(...).typed` step so the code can
be shared between `call` and `redispatch`.

[ghstack-poisoned]

peterbell10 committed:
Update on "Shard Operators.cpp"


This file can take 5 minutes on its own to compile, and is the single limiting
factor for compile time of `libtorch_cpu` on a 32-core threadripper. Instead,
sharding into 5 files that take around 1 minute each cuts a full minute off the
overall build time.

This also factors out the `.findSchemaOrThrow(...).typed` step so the code can
be shared between `call` and `redispatch`.

[ghstack-poisoned]

peterbell10 commented on tools/code_analyzer/op_deps_pass.cpp:
```diff
@@ -770,6 +770,12 @@ class OpDependency : public ModulePass {
+      // UndefValue inherits from ConstantValue, but don't contain any data
+      // See: https://llvm.org/docs/LangRef.html#undefined-values
+      return;
+    }
```
It turns out clang was generating an IR node type that it wasn't before, and it was causing infinite recursion.

peterbell10 committed:
Update on "Shard Operators.cpp"


This file can take 5 minutes on its own to compile, and is the single limiting
factor for compile time of `libtorch_cpu` on a 32-core threadripper. Instead,
sharding into 5 files that take around 1 minute each cuts a full minute off the
overall build time.

This also factors out the `.findSchemaOrThrow(...).typed` step so the code can
be shared between `call` and `redispatch`.

[ghstack-poisoned]

albanD commented:
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29962049).

peterbell10 committed:
Update on "Shard Operators.cpp"


This file can take 5 minutes on its own to compile, and is the single limiting
factor for compile time of `libtorch_cpu` on a 32-core threadripper. Instead,
sharding into 5 files that take around 1 minute each cuts a full minute off the
overall build time.

This also factors out the `.findSchemaOrThrow(...).typed` step so the code can
be shared between `call` and `redispatch`.

Differential Revision: [D29962049](https://our.internmc.facebook.com/intern/diff/D29962049)

[ghstack-poisoned]

peterbell10 committed:
Update on "Shard Operators.cpp"


This file can take 5 minutes on its own to compile, and is the single limiting
factor for compile time of `libtorch_cpu` on a 32-core threadripper. Instead,
sharding into 5 files that take around 1 minute each cuts a full minute off the
overall build time.

This also factors out the `.findSchemaOrThrow(...).typed` step so the code can
be shared between `call` and `redispatch`.

Differential Revision: [D29962049](https://our.internmc.facebook.com/intern/diff/D29962049)

[ghstack-poisoned]

albanD commented:
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29962049).

peterbell10 committed:
Update on "Shard Operators.cpp"


This file can take 5 minutes on its own to compile, and is the single limiting
factor for compile time of `libtorch_cpu` on a 32-core threadripper. Instead,
sharding into 5 files that take around 1 minute each cuts a full minute off the
overall build time.

This also factors out the `.findSchemaOrThrow(...).typed` step so the code can
be shared between `call` and `redispatch`.

Differential Revision: [D29962049](https://our.internmc.facebook.com/intern/diff/D29962049)

[ghstack-poisoned]

albanD commented:
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29962049).

albanD commented:
Sorry for the delay but there are (too many) internal config files that hardcoded the path to these files that I need to fix manually. Shouldn't take long now :)
